### ISSUE TITLE
Support node allocatable and capacity managed by external controller

### ISSUE THREAD

**Issue by 44past4** [kubernetes/kubernetes#120833]
### What would you like to be added?

I would like to propose adding new kubelet options which would stop kubelet from detecting, calculating and updating information about CPU and memory capacity and allocatable resources. Instead those values would need to be set by a dedicated external controller which could use any logic to provide them. Kubelet in such a setup would be watching for node CPU and memory capacity and allocatable resources changes on an API node resource status and react accordingly when those change.

### Why is this needed?

Many open source and commercial virtualization solutions like [KVM](https://www.linux-kvm.org/page/CPUHotPlug) and VMWare support dynamic CPU and memory hotplugging without the need to restart a VM. With those solutions we would like to have a way to update node capacity and allocatable resources to match the current VM state without a need to restart kubelet.

Moreover in cases when some resources are removed we would like to update node resources values before the actual change in resources is effective to avoid race conditions and situations when for instance some recently scheduled pods would need to be killed just after node capacity has been reduced or situations where pods require some grace period before being killed but the resources available on the node has already been changed. Because of those relying on kubelet to retrospectively detect the node capacity changes by itself is not sufficient.

Other situations when kubelet may not be able to detect resources available on a node on its own is when there are some non containerized workloads running on the same host. In this situation kubelet cannot properly detect how much resources on a given host are actually available for running pods. This information could be obtained from some external source be a node agent or a controller responsible for running those non containerized workloads and communicated to kubelet by updating node allocatable resources status.

A different solution to a similar problem has been proposed in [KEP-3953: Dynamic node resize](https://github.com/kubernetes/enhancements/pull/3955). This proposal addresses some issues with the other proposal by providing more flexibility when it comes to logic used to detect node capacity and allocatable resources and timing when those should be changed.

**Comment by sftim** [comment]
Overall, this feels like it's an addition to the discussion for https://github.com/kubernetes/enhancements/issues/3953

**Comment by marquiz** [comment]
What if CRI would be the interface and CRI runtime would be responsible for detecting the available resources?

Just for ref in [KEP-4033](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4033-group-driver-detection-over-cri) we added support for detecting the cgroup driver over CRI. In the discussion we speculated on more dynamic configuration changes over a gRPC streaming interface in CRI

**Comment by Karthik-K-N** [comment]
> What if CRI would be the interface and CRI runtime would be responsible for detecting the available resources?
> 
> Just for ref in [KEP-4033](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4033-group-driver-detection-over-cri) we added support for detecting the cgroup driver over CRI. In the discussion we speculated on more dynamic configuration changes over a gRPC streaming interface in CRI

Yeah, we have not explored the possibilities of using CRI to get the real time data, Thanks for suggesting will explore on that way as well.

**Comment by ffromani** [comment]
@44past4 Hi! does this issue supercede https://github.com/kubernetes/kubernetes/issues/120832 ? Or do you want to pursue them both? If the latter, could you please clarify the overlaps/interactions?

**Comment by 44past4** [comment]
I believe that the [#120832](https://github.com/kubernetes/kubernetes/issues/120832) is a much simpler request which addresses different use cases mainly for testing where we have a host with a static size but for some reason we do not want to expose all of its resources as a Kubernetes node.
This proposal addresses different use cases where the resources available on a given host can change and we need to make kubelet and other components aware of this change.
Taking into account that those are different use cases I believe that it makes sense to have both issues.

**Comment by ffromani** [comment]
/triage accepted

because related (maybe additional use case?) to https://github.com/kubernetes/enhancements/issues/3953
